File: sortings/merg_sort.cpp
// Merge Sort algorithm by MrPoJ 10th oct 2020
#include<bits/stdc++.h>
using namespace std;

const int maxN = 1e5+11;
int n, a[maxN],cnt;

void merge (int l, int r) {
        int mid = l + r >> 1;
        int p1 = l, p2 = mid+1, k = 0;
        int arr[n] = {};
        for (int i = l; i <= r; i++)
                if (p1 > mid)
                        arr[k++] = a[p2++];
                else if (p2 > r)
                        arr[k++] = a[p1++];
                else if (a[p1] < a[p2])
                        arr[k++] = a[p1++];
                else
                        arr[k++] = a[p2++];
        for (int i = 0; i < k; i++)
                a[l++] = arr[i];

}



void _sort (int l = 0, int r = n-1) {
	int mid = l + r >> 1;
	if (l < r) 
		_sort(l, mid), _sort (mid+1, r);
	 merge (l, r);
	 cout << cnt++ << ": ";
	 for (int i = 0; i < n; i++) cout << a[i] << ' ';
	 cout << endl;
}

int main () {
	cin >> n;
	for (int i = 0; i < n; i++)
		cin >> a[i];
	_sort ();
	return 0;
}

File: sortings/merg_sort.py
# Merg Sort algorithm by MrPoj 10th oct 2020

def merge (a, st, mid, en):
    p = st, q = mid + 1
    n = len(a)
    arr = {}
    k = 0
    for i in range (st, en+1):
        if p > mid:
            arr[k+1] = a[q+1]
            k+=1
            q+=1
        elif q > en:
            arr[k+1] = a[p+1]
            k+=1
            p+=1
        elif a[p] < a[q]:
            arr[k+1] = a[p+1]
            k+=1
            p+=1
        else:
            arr[k+1] = a[q+1]
            k+=1
            q+=1
    for i in range (0, k):
        a[st+1] = arr[i]
        st+=1

def _sort (a, st, en):
    if st < en:
        mid = (st + en) // 2
        _sort(a, st, mid)
        _sort(a, mid+1, en)
    merge (a, st, mid, en)

a = list(map(int, input().split()))
n = len(a)
_sort (a, 0, n)
print(a)
